### DISCUSSION TITLE
Elevation transforms and HAE vs MSL

### DISCUSSION THREAD

**Discussion by zdposu** [pelicanmapping/osgearth#1987] [Support Q&A]
I'm trying to figure out how best to consistently get HAE and MSL.  I wrote some test code from some previous forum discussions and I'm not getting the results I anticipated.  Can you guide me on how to get MSL and HAE elevations and, if possible, how to get the layer from which the elevation value was retrieved from?

Note that GeoPoint::transform doesn't seem to transform the elevation from one vdatum to another.  I'm also confused on why the last two VerticalDatum::transform calls always come back with the same results.

Any insight on what I'm doing wrong would be greatly appreciated!

Here is the sample code, the outputs, and the .earth file I am using:

**Inputs:**
	const std::vector< std::pair<double, double> > LAT_LON_LIST_UTAH
	{
		{40.5745329750, -111.7853343571 }, //LCC
		{40.5726146663, -111.7765861811 }, //Little Cottonwood Road fork
		{40.5888667984, -111.6388564531 }, //Alta
		{40.5817361188, -111.6569948744 }, //Snowbird
	    {40.3761752391, -105.5236120331 }, //Estes Park
		{45.4108516473, -75.6875127888 },  //Ottowa
		{38.9315317322, -119.9790876095}   //South Lake Tahoe
	};

**Test code:**

		const SpatialReference* mapSRS = mMapNode->getMapSRS();
		if (mapSRS->getVerticalDatum())
		{
			OE_INFO << mapSRS->getVerticalDatum()->getName() << std::endl;
		}
		else
		{
			OE_INFO << "geodetic (" << mapSRS->getEllipsoid()->getName() << ")" << std::endl;
		}

		auto srsToString = [](auto srs)
		{
			std::string retval = srs->getName();
			retval += ":";
			retval += srs->getVerticalDatum() ? srs->getVerticalDatum()->getName() : "geodetic";
			retval += "-";
			retval += srs->getEllipsoid()->getName();
			return retval;
		};
		
		auto terrain_srs = mMapNode->getTerrain()->getSRS();
		osgEarth::GeoPoint mapPoint(mMapNode->getTerrain()->getSRS(), aLatLon.y(), aLatLon.x());
		auto mapPoint_sample = mMap->getElevationPool()->getSample(mapPoint, &mElevWorkingSet);
		mapPoint.z() = mapPoint_sample.elevation().as(Units::METERS);

		OE_INFO << "!^!^!^! mapPoint elev(m): " << mapPoint_sample.elevation().as(Units::METERS) << std::endl;

		{
			auto srs_wgs84_msl = SpatialReference::create("wgs84", "egm96");
			GeoPoint mapPointEGM96 = mapPoint.transform(srs_wgs84_msl);
			OE_INFO << "!^!^!^! mapPointEGM96(m): " << mapPointEGM96.z() << " (" << srsToString(srs_wgs84_msl) << ")" << std::endl;
		}
		{
			auto srs_wgs84_egm84 = SpatialReference::create("wgs84", "egm84");
			GeoPoint mapPointEGM84 = mapPoint.transform(srs_wgs84_egm84);
			OE_INFO << "!^!^!^! mapPointEGM84(m): " << mapPointEGM84.z() << " (" << srsToString(srs_wgs84_egm84) << ")" << std::endl;
		}
		{
			auto srs_wgs84_hae = mMapNode->getMapSRS()->getGeodeticSRS();
			auto isGeodeticSRS = srs_wgs84_hae->isGeodetic();
			GeoPoint mapPointHAE = mapPoint.transform(srs_wgs84_hae);
			//GeoPoint mapPointGeodetic(mMapNode->getMapSRS()->getGeodeticSRS(), mapPoint);
			OE_INFO << "!^!^!^! mapPointHAE  (m): " << mapPointHAE.z() << " (" << srsToString(srs_wgs84_hae) << ")" << std::endl;
		}
		{
			auto srs_wgs84_egm08 = SpatialReference::create("wgs84", "egm2008");
			GeoPoint mapPointEGM08 = mapPoint.transform(srs_wgs84_egm08);
			OE_INFO << "!^!^!^! mapPointEGM96(m): " << mapPointEGM08.z() << " (" << srsToString(srs_wgs84_egm08) << ")" << std::endl;
		}

		if (true)
		{
			double egm96z = mapPoint.z();

			osgEarth::VerticalDatum::transform(
				mapPoint.getSRS()->getVerticalDatum(),
				osgEarth::VerticalDatum::get("egm96"),
				mapPoint.y(),
				mapPoint.x(),
				egm96z);
			OE_INFO << "!^!^!^! EGM96: " << egm96z << std::endl;

			double egm84z = mapPoint.z();

			osgEarth::VerticalDatum::transform(
				mapPoint.getSRS()->getVerticalDatum(),
				osgEarth::VerticalDatum::get("egm84"),
				mapPoint.y(),
				mapPoint.x(),
				egm84z);
			OE_INFO << "!^!^!^! EGM84: " << egm84z << std::endl;

			double egm08z = mapPoint.z();

			osgEarth::VerticalDatum::transform(
				mapPoint.getSRS()->getVerticalDatum(),
				osgEarth::VerticalDatum::get("egm2008"),
				mapPoint.y(),
				mapPoint.x(),
				egm84z);
			OE_INFO << "!^!^!^! EGM08: " << egm08z << std::endl;

			double geoidz = mapPoint.z();

			osgEarth::VerticalDatum::transform(
				mapPoint.getSRS()->getVerticalDatum(),
				mMapNode->getMapSRS()->getGeodeticSRS()->getVerticalDatum(),
				mapPoint.y(),
				mapPoint.x(),
				geoidz);
			OE_INFO << "!^!^!^! GEOID: " << geoidz << std::endl;
		}

**Output:**
[osgEarth]  geodetic (wgs 84)
[osgEarth]  !^!^!^! mapPoint elev(m): 1614.74
[osgEarth]  !^!^!^! mapPointEGM96(m): 1614.74 (WGS84:EGM96-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM84(m): 1614.74 (WGS84:EGM84-wgs 84)
[osgEarth]  !^!^!^! mapPointHAE  (m): 1614.74 (WGS84:geodetic-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM96(m): 1614.74 (WGS84:EGM2008-wgs 84)
[osgEarth]  !^!^!^! EGM96: 1631.77
[osgEarth]  !^!^!^! EGM84: 1629.87
[osgEarth]  !^!^!^! EGM08: 1614.74
[osgEarth]  !^!^!^! GEOID: 1614.74
[osgEarth]  geodetic (wgs 84)
[osgEarth]  !^!^!^! mapPoint elev(m): 1649.66
[osgEarth]  !^!^!^! mapPointEGM96(m): 1649.66 (WGS84:EGM96-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM84(m): 1649.66 (WGS84:EGM84-wgs 84)
[osgEarth]  !^!^!^! mapPointHAE  (m): 1649.66 (WGS84:geodetic-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM96(m): 1649.66 (WGS84:EGM2008-wgs 84)
[osgEarth]  !^!^!^! EGM96: 1666.68
[osgEarth]  !^!^!^! EGM84: 1664.78
[osgEarth]  !^!^!^! EGM08: 1649.66
[osgEarth]  !^!^!^! GEOID: 1649.66
[osgEarth]  geodetic (wgs 84)
[osgEarth]  !^!^!^! mapPoint elev(m): 2609.62
[osgEarth]  !^!^!^! mapPointEGM96(m): 2609.62 (WGS84:EGM96-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM84(m): 2609.62 (WGS84:EGM84-wgs 84)
[osgEarth]  !^!^!^! mapPointHAE  (m): 2609.62 (WGS84:geodetic-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM96(m): 2609.62 (WGS84:EGM2008-wgs 84)
[osgEarth]  !^!^!^! EGM96: 2626.31
[osgEarth]  !^!^!^! EGM84: 2624.47
[osgEarth]  !^!^!^! EGM08: 2609.62
[osgEarth]  !^!^!^! GEOID: 2609.62
[osgEarth]  geodetic (wgs 84)
[osgEarth]  !^!^!^! mapPoint elev(m): 2484.32
[osgEarth]  !^!^!^! mapPointEGM96(m): 2484.32 (WGS84:EGM96-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM84(m): 2484.32 (WGS84:EGM84-wgs 84)
[osgEarth]  !^!^!^! mapPointHAE  (m): 2484.32 (WGS84:geodetic-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM96(m): 2484.32 (WGS84:EGM2008-wgs 84)
[osgEarth]  !^!^!^! EGM96: 2501.07
[osgEarth]  !^!^!^! EGM84: 2499.21
[osgEarth]  !^!^!^! EGM08: 2484.32
[osgEarth]  !^!^!^! GEOID: 2484.32
[osgEarth]  geodetic (wgs 84)
[osgEarth]  !^!^!^! mapPoint elev(m): 2304.33
[osgEarth]  !^!^!^! mapPointEGM96(m): 2304.33 (WGS84:EGM96-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM84(m): 2304.33 (WGS84:EGM84-wgs 84)
[osgEarth]  !^!^!^! mapPointHAE  (m): 2304.33 (WGS84:geodetic-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM96(m): 2304.33 (WGS84:EGM2008-wgs 84)
[osgEarth]  !^!^!^! EGM96: 2317.47
[osgEarth]  !^!^!^! EGM84: 2316.54
[osgEarth]  !^!^!^! EGM08: 2304.33
[osgEarth]  !^!^!^! GEOID: 2304.33
[osgEarth]  geodetic (wgs 84)
[osgEarth]  !^!^!^! mapPoint elev(m): 66.3461
[osgEarth]  !^!^!^! mapPointEGM96(m): 66.3461 (WGS84:EGM96-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM84(m): 66.3461 (WGS84:EGM84-wgs 84)
[osgEarth]  !^!^!^! mapPointHAE  (m): 66.3461 (WGS84:geodetic-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM96(m): 66.3461 (WGS84:EGM2008-wgs 84)
[osgEarth]  !^!^!^! EGM96: 99.9251
[osgEarth]  !^!^!^! EGM84: 99.1827
[osgEarth]  !^!^!^! EGM08: 66.3461
[osgEarth]  !^!^!^! GEOID: 66.3461
[osgEarth]  geodetic (wgs 84)
[osgEarth]  !^!^!^! mapPoint elev(m): 1905.18
[osgEarth]  !^!^!^! mapPointEGM96(m): 1905.18 (WGS84:EGM96-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM84(m): 1905.18 (WGS84:EGM84-wgs 84)
[osgEarth]  !^!^!^! mapPointHAE  (m): 1905.18 (WGS84:geodetic-wgs 84)
[osgEarth]  !^!^!^! mapPointEGM96(m): 1905.18 (WGS84:EGM2008-wgs 84)
[osgEarth]  !^!^!^! EGM96: 1928.94
[osgEarth]  !^!^!^! EGM84: 1928.02
[osgEarth]  !^!^!^! EGM08: 1905.18
[osgEarth]  !^!^!^! GEOID: 1905.18

**Earth file:**

<map name="SEPTRE" type="geocentric" version="2">
        
	<options>
        <cache type="filesystem">
            <path>C:\temp\osgEarth\Cache</path>
        </cache>
		<profile>
            <srs>wgs84</srs>
            
        </profile>
    </options>

    <GDALImage name="Base World Imagery">
        <url>./world.tif</url>
        <cache_policy usage="none"/>
    </GDALImage>

    <ArcGISServerImage name="ArcGIS World Imagery">
        <url>http://services.arcgisonline.com/arcgis/rest/services/World_Imagery/MapServer/</url>
    </ArcGISServerImage>

    <XYZImage name="OSM Imagery" opacity="0.3">
        <url>http://[abc].tile.openstreetmap.org/{z}/{x}/{y}.png</url>
        <profile>spherical-mercator</profile>
        <cache_policy usage="none"/>
        <attribution>&#169;OpenStreetMap contributors</attribution>
    </XYZImage>

    <TMSElevation name="ReadyMap 90m Elevation">
        <url>http://readymap.org/readymap/tiles/1.0.0/116/</url>
        
    </TMSElevation>
	
</map>

**Comment by gwaldron** [comment]
Do me a favor - this is too much to digest and parse as-is. Please give us one example of one `GeoPoint` transformation you feel is in error, and the expected result, and we will go from there.

**Reply by nebsar** [reply]
I might have the similar problem, placing my aircraft camera on the earth with 0 elevation. I am not sure what the problem is.

**Comment by zdposu** [comment]
Here is a more concise example:

		auto srsToString = [](auto srs)
		{
			std::string retval = srs->getName();
			retval += ":";
			retval += srs->getVerticalDatum() ? srs->getVerticalDatum()->getName() : "geodetic";
			retval += "-";
			retval += srs->getEllipsoid()->getName();
			return retval;
		};
		
		auto terrain_srs = mMapNode->getTerrain()->getSRS();
		osgEarth::GeoPoint mapPoint(mMapNode->getTerrain()->getSRS(), aLatLon.y(), aLatLon.x());
		auto mapPoint_sample = mMap->getElevationPool()->getSample(mapPoint, &mElevWorkingSet);
		mapPoint.z() = mapPoint_sample.elevation().as(Units::METERS);

		OE_INFO << "!^!^!^! mapPoint elev(m): " << mapPoint_sample.elevation().as(Units::METERS) << std::endl;

		{
			auto srs_wgs84_msl = SpatialReference::create("wgs84", "egm96");
			GeoPoint mapPointEGM96 = mapPoint.transform(srs_wgs84_msl);
			OE_INFO << "!^!^!^! mapPointEGM96(m): " << mapPointEGM96.z() << " (" << srsToString(srs_wgs84_msl) << ")" << std::endl;
		}
		{
			auto srs_wgs84_hae = mMapNode->getMapSRS()->getGeodeticSRS();
			auto isGeodeticSRS = srs_wgs84_hae->isGeodetic();
			GeoPoint mapPointHAE = mapPoint.transform(srs_wgs84_hae);
			OE_INFO << "!^!^!^! mapPointHAE  (m): " << mapPointHAE.z() << " (" << srsToString(srs_wgs84_hae) << ")" << std::endl;
		}

Gives the output (with input LL 40.5745329750, -111.7853343571):

[osgEarth]  !^!^!^! mapPoint elev(m): 1597.71
[osgEarth]  !^!^!^! mapPointEGM96(m): 1597.71 (WGS84:EGM96-wgs 84)
[osgEarth]  !^!^!^! mapPointHAE  (m): 1597.71 (WGS84:geodetic-wgs 84)

Basically, I would expect these elevation values to differ.  The goal is to get MSL and HAE from the same elevation pool sample by using GeoPoint::transform with different SRS/vdatums.
